zerver/data_import/mattermost.py

```diff
@@ -0,0 +1,540 @@
+import os
+import logging
+import subprocess
+import ujson
+import re
+
+from typing import Any, Callable, Dict, List, Set
+
+from django.conf import settings
+from django.utils.timezone import now as timezone_now
+
+from zerver.models import Recipient
+from zerver.lib.utils import (
+    process_list_in_batches,
+)
+from zerver.data_import.import_util import ZerverFieldsT, build_zerver_realm, \
+    build_stream, build_realm, build_message, \
+    create_converted_data_files, make_subscriber_map, build_recipients, \
+    build_user_profile, build_stream_subscriptions, \
+    build_personal_subscriptions, build_user_message, SubscriberHandler
+
+from zerver.data_import.mattermost_user import UserHandler
+from zerver.data_import.sequencer import NEXT_ID, IdMapper
+
+def make_realm(realm_id: int, team: Dict[str, Any]) -> ZerverFieldsT:
+    # set correct realm details
+    NOW = float(timezone_now().timestamp())
+    domain_name = settings.EXTERNAL_HOST
+    realm_subdomain = team["name"]
+
+    zerver_realm = build_zerver_realm(realm_id, realm_subdomain, NOW, 'Mattermost')
+    realm = build_realm(zerver_realm, realm_id, domain_name)
+
+    # We may override these later.
+    realm['zerver_defaultstream'] = []
+
+    return realm
+
+def process_user(user_dict: Dict[str, Any], realm_id: int, team_name: str,
+                 user_handler: UserHandler, user_id_mapper: IdMapper) -> ZerverFieldsT:
+    def is_team_admin(user_dict: Dict[str, Any]) -> bool:
+        for team in user_dict["teams"]:
+            if team["name"] == team_name and "team_admin" in team["roles"]:
+                return True
+        return False
```

Comment:
We probably want this function to accept the team that we're importing as an argument, as a user could be an admin in one team but not others, and we want import to preserve that.

Reply:
Also see below that maybe we should set this up in the same code path that sets up mirror dummy.

Reply:
Actually in mirror dummy setup we only iterate through messages in a realm and set up mirror dummy user the sender is not in realm. So we are only covering the users who have sent at least one message.